runtime/common/tests/xcm_tests.rs

```diff
@@ -2156,3 +2157,579 @@ fn less_than_min_xcm_fee_should_not_work() {
         ));
     });
 }
+
+#[test]
+fn transfer_multicurrencies_should_work_scenarios() {
+    MockNet::reset();
+
+    let para_a_id = 1;
+    let para_b_id = 2;
+    let para_a_source_location = AssetLocation(VersionedMultiLocation::V1(MultiLocation::new(
+        1,
+        X1(Parachain(para_a_id)),
+    )));
+    let para_b_source_location = AssetLocation(VersionedMultiLocation::V1(MultiLocation::new(
+        1,
+        X1(Parachain(para_b_id)),
+    )));
+    let units_per_sec = 0;
+
+    let para_a_asset_metadata =
+        create_asset_metadata("ParaAToken", "ParaA", 18, 1, None, false, true);
+    let para_b_asset_metadata =
+        create_asset_metadata("ParaBToken", "ParaB", 18, 1, None, false, true);
+
+    let relay_source_location = AssetLocation(VersionedMultiLocation::V1(MultiLocation::parent()));
+    let relay_asset_metadata = create_asset_metadata("Kusama", "KSM", 12, 1, None, false, true);
+
+    // Register ParaA native asset in ParaA
+    let _ = register_assets_on_parachain::<ParaA>(
+        &para_a_source_location,
+        &para_a_asset_metadata,
+        Some(units_per_sec),
+        None,
+    );
+    // Register relaychain native asset in ParaA
+    let relay_asset_id_on_a = register_assets_on_parachain::<ParaA>(
+        &relay_source_location,
+        &relay_asset_metadata,
+        Some(units_per_sec),
+        None,
+    );
+    // register ParaB native asset on ParaA
+    let b_asset_id_on_a = register_assets_on_parachain::<ParaA>(
+        &para_b_source_location,
+        &para_b_asset_metadata,
+        Some(units_per_sec),
+        None,
+    );
+
+    // register ParaB native asset on ParaB
+    let b_asset_id_on_b = register_assets_on_parachain::<ParaB>(
+        &para_b_source_location,
```

Comment:
how come the locations are identical between paraA and B? I thought XCM locations are relative to the source?

Reply:
Yes you're right, the correct way of representing a self location (native asset location) should be (0, Here). 
However representing in a "absolute" way (1, X1(self_para_id)) was a common mistake that many teams copied, including us. 
We worked around this mistake `impl<T, Balance> MatchesFungible<Balance> for IsNativeConcrete<T>` where we also check for `MultiLocation::is_here` 
Actually my next PR will be to actually start using the correct way of handling self locations and the code is ready, just need the other PRs to go in first. https://github.com/Manta-Network/Manta/issues/452